### PR TITLE
folly: support macOS

### PR DESCRIPTION
**Pull request by strager**
###### Motivation for this change

Enable building the folly library on macOS (Darwin).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by veprbl** [comment]
The patches to disable tests might require more work to maintain. Do you think we could do it cleaner with just `substituteInPlace`?

**Comment by strager** [comment]
@veprbl, I uploaded a new version which uses substituteInPlace. I don't have experience maintaining packages to know which is better. Let me know what you think about the substituteInPlace version compared to the old version (https://github.com/strager/nixpkgs/commit/700f99dd6236d8caba357569a2e265f53523c16d).

**Review by veprbl — APPROVED** [review]
LGTM

**Comment by veprbl** [comment]
@GrahamcOfBorg build folly

**Comment by veprbl** [comment]
Fails to build

**Comment by strager** [comment]
Yeah, f75c68c4490057a35d04726949d67dd5b092fc9f introduced problems. I'll rebase and fix them.

File: pkgs/development/libraries/folly/macos-10.11.patch [hunk 1]
```diff
@@ -0,0 +1,88 @@
+--- a/folly/IndexedMemPool.h
```
Comment (veprbl) [inline pkgs/development/libraries/folly/macos-10.11.patch]: I can build folly without this patch. Is this for uses in the code built with clang 5?

**Review by veprbl — COMMENTED** [review]
Since we just had a breakage due to the patches, perhaps we could remove some of them.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -10167,7 +10167,10 @@ in
 
   fontconfig-ultimate = callPackage ../development/libraries/fontconfig-ultimate {};
 
-  folly = callPackage ../development/libraries/folly { };
+  folly = callPackage ../development/libraries/folly {
+    # Clang 5 miscompiles folly. Build with a newer compiler on Darwin.
+    stdenv = if stdenv.isDarwin then llvmPackages_7.stdenv else stdenv;
```
Comment (veprbl) [inline pkgs/top-level/all-packages.nix]: Not needed anymore since ae1e940cc0c4767adba1ea8f4a25bda569be26c5 has just landed on master #60491

**Comment by veprbl** [comment]
macOS support was merged as a part of #62330

**Review by matthewbauer — APPROVED** [review]

**Comment by veprbl** [comment]
@strager Are you still interested in contributing the tests?

**Comment by strager** [comment]
> @strager Are you still interested in contributing the tests?

I have no plans to work on this PR or on folly. You can take over my work if you want.

**Comment by SuperSandro2000** [comment]
@strager please resolve the merge conflicts.

**Comment by prusnak** [comment]
Closing. macOS support was added #62330 and @stragger expressed he does not want to work on the tests: https://github.com/NixOS/nixpkgs/pull/59846#issuecomment-569870158

**Pull request closed without merging**